src/DataTransferObject.php

```diff
@@ -35,13 +48,20 @@ public function __construct(...$args)
             throw UnknownProperties::new(static::class, array_keys($args));
         }
 
+        return $this;
+    }
+
+    protected function validate(DataTransferObjectClass $class): static
```

Comment:
Do you think there may be value to making the validate method public? For example, the new method would set the properties on the DTO, and then the validate method would be called.

It's probably a good idea to put the validation logic in its own resolver or something as well.

```php
public function validate(): static
{
    //
}
```

Reply:
On this note, if we have validation providers, all the logic for validating DTOs could be handled by a `DataTransferObjectValidationProvider` that is attributed to the base class (just like the caster).

For example:

```php
#[ValidateWith(DataTransferObjectValidationProvider::class)]
abstract class DataTransferObject
{
    //
}
```

Reply:
> I think we will need to make the validate method public for it to work with sub-DTOs that need to be validated

Can't you call protected methods from sub DTOs? I thought that was possible but haven't tested it.

Reply:
About the validation provider: can you refresh my memory on when it would be useful to override it? What I'm questioning here is whether it needs to be as configurable as you describe?

Reply:
> Can't you call protected methods from sub DTOs? I thought that was possible but haven't tested it.

By "sub DTOs" I am referencing data transfer object properties that are cast using `newWithoutValidation` and then need to be validated during the validation step (so as to roll up all the errors into a single exception).

I am envisioning a call like:

```php
PersonData::newWithoutValidation(firstName: 'Aidan', lastName: 'Casey')->validate();
```

> About the validation provider: can you refresh my memory on when it would be useful to override it? What I'm questioning here is whether it needs to be as configurable as you describe?

Right now rules can be created at a property level, but it would be awesome if a provider could handle a whole class as well. This provider would be passed an instance of the `ClassDescriptor` which would allow it to intimately validate the properties. For example, someone (a.k.a. - me  👀 ) could create a LaravelValidationProvider which uses attributes as rules, but the provider handles reading those attributes and making assertions based on them.

This functionality could be utilized for validating the sub DTOs as well by checking that the property type extends a DTO and calling the `validate` method on it as well. Similar to the `DataTransferObjectCaster` but for validation.